crates/wasmtime/src/component/values.rs

```diff
@@ -0,0 +1,389 @@
+            Val::Flags(Flags { ty, .. }) => Type::Flags(ty.clone()),
+        }
+    }
+
+    /// Serialize this value as core Wasm stack values.
+    pub(crate) fn lower<T>(
+        &self,
+        store: &mut StoreContextMut<T>,
+        options: &Options,
+        dst: &mut MaybeUninit<[ValRaw; MAX_STACK_PARAMS]>,
+        offset: usize,
+    ) -> Result<()> {
+        match self {
+            Val::Unit => (),
+            Val::Bool(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(if *value { 1 } else { 0 }));
+            }
+            Val::S8(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i32(*value as i32));
+            }
+            Val::U8(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(*value as u32));
+            }
+            Val::S16(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i32(*value as i32));
+            }
+            Val::U16(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(*value as u32));
+            }
+            Val::S32(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i32(*value));
+            }
+            Val::U32(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(*value));
+            }
+            Val::S64(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i64(*value));
+            }
+            Val::U64(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u64(*value));
+            }
+            Val::Float32(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::f32(*value));
+            }
+            Val::Float64(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::f64(*value));
+            }
+            Val::Char(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(u32::from(*value)));
+            }
+            Val::String(value) => {
+                let (ptr, len) = super::lower_string(
+                    &mut MemoryMut::new(store.as_context_mut(), options),
+                    value,
+                )?;
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i64(ptr as i64));
+                map_maybe_uninit!(dst[offset + 1]).write(ValRaw::i64(len as i64));
+            }
+            Val::List(List { values, .. }) => {
+                let (ptr, len) =
+                    lower_list(&mut MemoryMut::new(store.as_context_mut(), options), values)?;
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i64(ptr as i64));
+                map_maybe_uninit!(dst[offset + 1]).write(ValRaw::i64(len as i64));
+            }
+            Val::Record(Record { values, .. }) | Val::Tuple(Tuple { values, .. }) => {
+                for (index, value) in values.iter().enumerate() {
+                    value.lower(store, options, dst, offset + index)?;
```

Comment:
I think this `+ index` may not be right since each of the `values` may take up more than one `ValRaw` slot.

It might make more sense for this function to take `&mut std::slice::Iter<'_, MaybeUninit<ValRaw>>` instead of the slice/offset it currently has perhaps?

Reply:
One nice part of perhaps using `MaybeUninit<ValRaw>` is that you could delegate to the `Lower for T` implementations? Or otherwise I think it would be worthwhile figuring out how to delegate to that for primitives and strings/lists perhaps (may require an unsafe method or two to extract `MaybeUninit<[ValRaw; 2]>` from `&mut std::slice::Iter<'_, MaybeUninit<ValRaw>>` though.

Reply:
@alexcrichton Is there a convenient way to make a `&mut std::slice::Iter<'_, &mut MaybeUninit<ValRaw>>` from a `&mut MaybeUninit<[ValRaw; N]`?  I could do something like `(0..N).map(|i| map_maybe_uninit!(dst[i])`, but I don't think that's what you meant.

Reply:
Not easily, but you can transmute `&mut MaybeUninit<[ValRaw; N]>` to `&mut [MaybeUninit<ValRaw>; N]` "safely" and then `.iter()` from there